### PR TITLE
Utilize I/O vectors for outgoing distribution data

### PR DESCRIPTION
**Pull request by rickard-green**
This modification of the serialization of Erlang data to the external term format utilize an I/O vector instead of encoding to one continues binary. The major benefit is that it makes it possible to refer to off-heap binaries directly instead of copying the binary data into the resulting external term format. This I/O vector is passed to the distribution controller port in a port command, or to the distribution controller process via [`erlang:dist_ctrl_get_data()`](http://erlang.org/~rickard/OTP-15618/erts-10.4/doc/html/erlang.html#dist_ctrl_get_data-1).

Another large benefit is that it simplifies "hopefull encoding" when we do not know the capabilities of the receiver. This occurs when we asynchronously connect to a new node. We encode the term according to a predifined set of capabilities that is likely that the receiver will have. If the receiver do not have these capabilities we have to change the encoding according to actual capabilities. When encoding as an I/O vector it is easy to mark places that might have to be adjusted and also easy to adjust the content according to the actual capabilites of the receiver.

#### Return Type of erlang:dist_ctrl_get_data() ####

[`erlang:dist_ctrl_get_data()`](http://erlang.org/~rickard/OTP-15618/erts-10.4/doc/html/erlang.html#dist_ctrl_get_data-1) is documented to return data of type `iodata()`. Some users might want the result to be guaranteed to be of the [`iovec()`](http://erlang.org/~rickard/OTP-15618/erts-10.4/doc/html/erlang.html#type-iovec) type. For example, OTPs own implementation of the distribution over TLS. In the current version of this pull request this can be guaranteed by setting the `get_iovec` option in a call to [erlang:dist_ctrl_set_opt/3](http://erlang.org/~rickard/OTP-15618/erts-10.4/doc/html/erlang.html#dist_ctrl_set_opt-3). Another solution could be to restrict the return type of `erlang:dist_ctrl_get_data()` to [`iovec()`](http://erlang.org/~rickard/OTP-15618/erts-10.4/doc/html/erlang.html#type-iovec). This is backwards compatible since all `iovec()` data also is valid `iodata()` data and we do not see any major future improvements that could benefit from the return type being kept as `iodata()`. It is however also hard to foresee future changes that we might want to do, and if we change the return type to `iovec()` it might prevent future improvements that we might want to do.

#### erlang:term_to_iodata() ####

Besides the above, serialization of an Erlang term can also be done via [`term_to_binary()`](http://erlang.org/~rickard/OTP-15618/erts-10.4/doc/html/erlang.html#term_to_binary-2).

In order to provide the same functionality also for users of `term_to_binary()` this pull request introduce a new BIF [`term_to_iodata()`](http://erlang.org/~rickard/OTP-15618/erts-10.4/doc/html/erlang.html#term_to_iodata-2). The `term_to_iodata()` function will often return data that is not valid
[`iovec()`](http://erlang.org/~rickard/OTP-15618/erts-10.4/doc/html/erlang.html#type-iovec) data (but valid `iodata()`). In this case it will optimize for heap size/virtual binary heap size usage. If one pass the `iovec` option to `term_to_iodata/2` the return type will be `iovec()`. The `iovec` option optimize for passing the data to native code (drivers/NIFs).

Here there are lots of different solutions that could be chosen besides introduction of `term_to_iodata()` as in this pull request.

*   Instead of introducing `term_to_iodata/2` one could introduce `iodata` and `iovec` options to `term_to_binary/2`. It however feels a bit strange that `term_to_binary/2` in this case not necessarily return a binary.
*   Instead of introducing a `iovec` option to `term_to_iodata()` one could introduce a `term_to_iovec()`.
*   BIFs accepting `iodata()` as input are usually called something with iolist. For example, `iolist_to_binary()` which really should have been called `iodata_to_binary()`. By calling the new BIF `term_to_iolist()` instead of `term_to_iodata()` one would be consistent with how naming has been done before, but then we perhaps should go for new options to `term_to_binary()` since the name anyhow wont reflect what is actually returned.

### CODE REVIEW

**Comment by essen** [comment]
We are about to test this with RabbitMQ. About the questions:

I would be fine with `erlang:dist_ctrl_get_data()` simply returning `iovec()` but leaving the spec as `iodata()` for the time being. This would be easier to stay compatible with multiple Erlang versions, which we need, since we could just do some conversion to iovec for older releases that still only do iodata.

Maybe name the function `term_to_external` with options `binary`, `iovec` and `iodata`, since it's called the external term format after all. Then you could also have an `external_to_term` that accepts more than just binaries.

**Comment by rickard-green** [comment]
> I would be fine with `erlang:dist_ctrl_get_data()` simply returning `iovec()` but leaving the spec as `iodata()` for the time being. This would be easier to stay compatible with multiple Erlang versions, which we need, since we could just do some conversion to iovec for older releases that still only do iodata.

I'm not sure I follow. I don't see the compatibility issue since `iovec()` is valid `iodata()`. If you want your code to be able to run on both old an new releases it just have to handle it as `iodata()`, or am I missing something?

> Maybe name the function `term_to_external` with options `binary`, `iovec` and `iodata`, since it's called the external term format after all. Then you could also have an `external_to_term` that accepts more than just binaries.

Thanks! We will take this option into consideration when deciding on the final API.

**Comment by essen** [comment]
> I'm not sure I follow. I don't see the compatibility issue since `iovec()` is valid `iodata()`. If you want your code to be able to run on both old an new releases it just has to handle it as `iodata()`, or am I missing something?

I made a few assumptions here. Sometimes when you restrict specs Dialyzer ends up complaining if you still try to handle the old type. For example if you change a type from `map() | list()` to just `map()` it'll complain if you try to use it as a list to support older versions. I'm not completely sure it'd apply in this case and it's likely to be just a minor annoyance at most.

**Comment by rickard-green** [comment]
> I made a few assumptions here. Sometimes when you restrict specs Dialyzer ends up complaining if you still try to handle the old type. For example if you change a type from `map() | list()` to just `map()` it'll complain if you try to use it as a list to support older versions. I'm not completely sure it'd apply in this case and it's likely to be just a minor annoyance at most.

Yes, dialyzer might complain, but I don't think there is a way around that if you ever are going to change the type, and I think that it at least eventually should be changed to the correct type.

I don't think dialyzer will complain about this specific change though, but that is just a guess.

**Comment by kjnilsson** [comment]
This is nice. So would it be beneficial to use `term_to_iodata` with the `iovec` option instead of `term_to_binary` when writing to a raw file handle as well?

**Comment by essen** [comment]
I've started a branch to use this in our LZ4-compressed dist driver: https://github.com/essen/inet_tcp_compress_dist/pull/1

It works but we have not done measurements yet.

Not directly related, but do you think something could be done about the input as well? Right now a `recv` call implies doing the following:

``` erlang
binary_to_list(iolist_to_binary(decompress(Context, iolist_to_binary(Data))))
```

Maybe the right-most `iolist_to_binary/1` can be removed by setting socket options to `binary` (I *think* it ought to work) but right now `dist_util` only works with lists so the conversion is necessary. My decompress function returns an `iovec()` already (for reasons unrelated to this PR) so if we could make `dist_util` work with `iovec()` as well that'd avoid any conversion (but would require an `iovec_to_term/1`). Alternatively one conversion to binary would still be better than nothing and adding support for `binary` to `dist_util` ought to be much simpler. Do you think such a patch would be accepted?

Thanks.

**Comment by essen** [comment]
I removed the unrelated part of my comment because I am just confusing pre/post handshake states and there's no problem like I mentioned post-handshake. Please ignore.

**Comment by garazdawi** [comment]
> So would it be beneficial to use `term_to_iodata` with the `iovec` option instead of `term_to_binary` when writing to a raw file handle as well?

yes.

**Comment by rickard-green** [comment]
OTP Technical Board decision:
* Restrict return type of `erlang:dist_ctrl_get_data()` to `iovec()`
* Introduce an `erlang:term_to_iovec()` BIF with return type `iovec()` behaving exactly as `erlang:term_to_binary()` with the exception of return type

**Comment by essen** [comment]
Thanks! We've started extensive testing with this branch today and have not had any issue so far. It's too early to know how much this improves things, though. We'll report back.

File: erts/emulator/beam/atom.names [hunk 1]
```diff
@@ -290,6 +290,7 @@ atom gc_minor_start
 atom Ge='>='
 atom generational
 atom get_all_trap
+atom get_iovec
```
Comment (essen) [inline erts/emulator/beam/atom.names:293]: This one is probably no longer needed.

**Pull request merged**